app/questionnaire/router.py

```diff
@@ -0,0 +1,52 @@
+class Router:
+
+    def __init__(self, schema, routing_path, completeness, current_location=None):
+        self._schema = schema
+        self._routing_path = routing_path
+        self._completeness = completeness
+        self._current_location = current_location
+
+    def can_access_location(self):
+        if self._is_valid_location() and not self._get_first_incomplete_block_in_section() and not self._get_first_incomplete_block_in_survey():
+            return True
+
+        return False
+
+    def get_next_location(self):
+        if not self._is_valid_location():
+            next_location = self._get_expected_location()
+            return next_location
+
+        return self._get_first_incomplete_block_in_section() or self._get_first_incomplete_block_in_survey() or self._get_expected_location()
+
+    def _is_valid_location(self):
+        return self._current_location in self._routing_path
+
+    def _get_expected_location(self):
+        next_location = self._routing_path[0]
+
+        if self._completeness.completed_blocks:
+            next_location = self._completeness.get_first_incomplete_location_in_survey()
+            if not next_location:
+                next_location = self._routing_path[-1]
+
+        return next_location
+
+    def _get_first_incomplete_block_in_section(self):
```

Comment:
This method name doesn't quite match the functionality because the method only returns the first incomplete block if they aren't allowed access the block they're trying to get to.

I don't know if it's much better but perhaps `_get_section_redirect_location`.

Again - needs docstrings. This stuff isn't obvious :)

Reply:
I changed the names so many times as there is not any great ones, ya ill go with your suggestion and add doc strings to help.